packages/framework-provider-local/src/services/read-model-registry.ts

```diff
@@ -32,27 +37,67 @@ export class ReadModelRegistry {
     const queryPromise = new Promise((resolve, reject) =>
       cursor.exec((err, docs) => {
         if (err) reject(err)
-        else resolve(docs)
+        else {
+          docs.map((doc: ReadModelEnvelope) => {
+            doc.value.boosterMetadata!.optimisticConcurrencyValue = this.generateOptimisticConcurrencyValue(doc.value)
+          })
+          resolve(docs)
+        }
       })
     )
-
     return queryPromise as Promise<Array<ReadModelEnvelope>>
   }
 
-  public async store(readModel: ReadModelEnvelope): Promise<void> {
+  public async store(readModel: ReadModelEnvelope, expectedCurrentVersion: number): Promise<void> {
+    const readModelOptimistic: ReadModelEnvelope & { uniqueKey?: string } = readModel
+    readModelOptimistic.uniqueKey = `${readModel.typeName}_${readModel.value.id}_${readModel.value.boosterMetadata?.version}`
+    if (readModelOptimistic.value.boosterMetadata?.version === 1) {
+      return this.insert(readModel)
+    }
+    return this.update(readModelOptimistic, expectedCurrentVersion)
+  }
+
+  private insert(readModel: ReadModelEnvelope): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.readModels.insert(readModel, (err: unknown) => {
+        err ? reject(err) : resolve()
+      })
+    })
+  }
+
+  private update(readModel: ReadModelEnvelope, expectedCurrentVersion: number): Promise<void> {
     return new Promise((resolve, reject) => {
       this.readModels.update(
-        //use nedb dot notation value.id to match the record (see https://github.com/louischatriot/nedb#finding-documents)
-        { typeName: readModel.typeName, 'value.id': readModel.value.id },
+        {
+          typeName: readModel.typeName,
+          'value.id': readModel.value.id,
+          $or: [
+            { 'value.boosterMetadata.optimisticConcurrencyValue': { $exists: false } },
+            { 'value.boosterMetadata.optimisticConcurrencyValue': expectedCurrentVersion },
+          ],
+        },
         readModel,
-        { upsert: true },
-        (err) => {
+        { upsert: false, returnUpdatedDocs: true },
+        (err: unknown, numAffected: number) => {
+          if (numAffected === 0) {
```

Comment:
I though when we have an error updating something, numAffected is always 0. How do you know that if this condition is true, then it must be a optimistic concurrency error?

Reply:
Yes, it's not a perfect solution but in this case we are updating read models only son we can expect that the updated documents will be greater than 0. In this case, using the query (version = expected version) while updating we will have 0 documents updated if the row was updated by another instance.